calico-vpp-agent/services/service_server.go

```diff
@@ -227,57 +260,50 @@ func NewServiceServer(vpp *vpplink.VppLink, k8sclient *kubernetes.Clientset, log
 		},
 	)
 
+	// ---- Watch EndpointSlices (IPv4 + IPv6) ----
 	endpointStore, endpointInformer := cache.NewInformerWithOptions(
 		cache.InformerOptions{
 			ListerWatcher: cache.NewListWatchFromClient(
-				k8sclient.CoreV1().RESTClient(),
-				"endpoints",
+				k8sclient.DiscoveryV1().RESTClient(),
+				"endpointslices",
 				"",
 				fields.Everything(),
 			),
-			ObjectType:   &v1.Endpoints{},
+			ObjectType:   &discoveryv1.EndpointSlice{},
 			ResyncPeriod: 60 * time.Second,
 			Handler: cache.ResourceEventHandlerFuncs{
 				AddFunc: func(obj interface{}) {
-					endpoints, ok := obj.(*v1.Endpoints)
+					slice, ok := obj.(*discoveryv1.EndpointSlice)
 					if !ok {
-						panic("wrong type for obj, not *v1.Endpoints")
+						panic("wrong type for obj, not *discoveryv1.EndpointSlice")
 					}
-					server.handleServiceEndpointEvent(
-						server.resolveLocalServiceFromEndpoints(endpoints),
-						nil,
-					)
+					local := server.resolveLocalServiceFromEndpointSlice(slice, true /* newSlice */)
+					server.handleServiceEndpointEvent(local, nil)
 				},
-				UpdateFunc: func(old interface{}, obj interface{}) {
-					endpoints, ok := obj.(*v1.Endpoints)
+				UpdateFunc: func(old, new interface{}) {
+					oldSlice, ok := old.(*discoveryv1.EndpointSlice)
 					if !ok {
-						panic("wrong type for obj, not *v1.Endpoints")
+						panic("wrong type for obj, not *discoveryv1.EndpointSlice")
 					}
-					oldEndpoints, ok := old.(*v1.Endpoints)
+					newSlice, ok := new.(*discoveryv1.EndpointSlice)
 					if !ok {
-						panic("wrong type for old, not *v1.Endpoints")
+						panic("wrong type for obj, not *discoveryv1.EndpointSlice")
 					}
-					server.handleServiceEndpointEvent(
-						server.resolveLocalServiceFromEndpoints(endpoints),
-						server.resolveLocalServiceFromEndpoints(oldEndpoints),
-					)
+					oldLocal := server.resolveLocalServiceFromEndpointSlice(oldSlice, false /* newSlice */)
+					newLocal := server.resolveLocalServiceFromEndpointSlice(newSlice, true /* newSlice */)
+					server.handleServiceEndpointEvent(newLocal, oldLocal)
 				},
 				DeleteFunc: func(obj interface{}) {
-					switch value := obj.(type) {
-					case cache.DeletedFinalStateUnknown:
-						endpoints, ok := value.Obj.(*v1.Endpoints)
+					slice, ok := obj.(*discoveryv1.EndpointSlice)
+					if !ok {
+						slice, ok = obj.(cache.DeletedFinalStateUnknown).Obj.(*discoveryv1.EndpointSlice)
 						if !ok {
-							panic(fmt.Sprintf("obj.(cache.DeletedFinalStateUnknown).Obj not a (*v1.Endpoints) %v", obj))
+							panic(fmt.Sprintf("obj.(cache.DeletedFinalStateUnknown).Obj not a (*v1.EndpointSlice) %v", obj))
 						}
-						server.deleteServiceByName(serviceID(&endpoints.ObjectMeta))
-					case *v1.Endpoints:
-						server.deleteServiceByName(serviceID(&value.ObjectMeta))
-					default:
-						log.Errorf("unknown type in endpoints deleteFunction %v", obj)
 					}
+					server.deleteServiceByName(serviceID(&slice.ObjectMeta))
```

Comment:
I don't think we can remove this here.
Probably better to call again :
```
server.handleServiceEndpointEvent(oldLocal, newLocal)
```

Also we'll need to get the name of the corresponding service being deleted, for the `` case, we'll need to keep a reference around to `endpointSlices` keyed by name so that we can find the trailing object, and fetch its name.